plutus-tx-plugin/src/Language/PlutusTx/Compiler/Builtins.hs

```diff
@@ -366,40 +395,27 @@ scottBoolToHaskellBool = do
         PIR.LamAbs () arg scottBoolTy $
         PIR.mkIterApp () instantiatedMatch [ haskellTrue, haskellFalse ]
 
--- | Wrap an integer relation of arity @n@ that produces a Scott boolean.
-wrapIntRel :: Compiling uni m => Int -> PIRTerm uni -> m (PIRTerm uni)
-wrapIntRel arity term = do
-    intTy <- lookupBuiltinType ''Integer
-    args <- replicateM arity $ do
+-- | Eta-expand a function with the given argument types.
+etaExpand :: Compiling uni m => [PIRType uni] -> PIRTerm uni -> m (PIRTerm uni)
+etaExpand argTys term = do
```

Comment:
That only works for monomorphic functions, right? `withTypedBuiltinName` gives you the meaning of a builtin, from which you can extract its `TypeScheme`, which you can use to generically eta-expand the builtin regardless of whether it's monomorphic or polymorphic.

Reply:
Yes, this is very simplistic, and for a known set of things. I may well take your suggestion if we start having more varied builtin functions lying around!

Reply:
Like `ifThenElse` that is going to land the next week?

Reply:
That's not going to be exposed by the Plutus Tx prelude, so doesn't need a binding like this to wire it up. 

Besides, in many cases we also need to do things like insert the adapter between builtin booleans and Haskell-compiled booleans (which will be `ifThenElse` in future), which has to go inside the eta-expansion... so I suspect this is always going to be fairly custom.

Reply:
> That's not going to be exposed by the Plutus Tx prelude, so doesn't need a binding like this to wire it up.

Right.

> Besides, in many cases we also need to do things like insert the adapter between builtin booleans and Haskell-compiled booleans

Haskell-compiled booleans are going to be embedded in Plutus like integers or bytestring in the same PR that adds `IfThenElse` (which is exactly the reason why we have to add `IfThenElse` as a builtin).

Reply:
> Haskell-compiled booleans are going to be embedded in Plutus like integers or bytestrings

*Haskell* booleans are going to be embedded in PLC as builtin types. I'm talking about Haskell-compiled booleans from the Plutus Tx compiler, i.e. the Scott-encoded, abstract datatypes that the rest of the program talks about.

Reply:
Oh, thanks. Would it make sense to tweak the compiler not to compile `Bool` to a Scott-encoded type, but instead to embedded `Bool`? It's going to be clunky if `Integer` and `ByteString` are compiled via `typeRepBuiltin` and `liftBuiltin`, but `Bool`, which is also in the universe, is not compiled the same way. Or am I again misunderstanding something?

Reply:
We already treat them differently, which is a bit clunky, but it's for a reason.

The difference is that `Bool` can just be treated Haskell datatype, and then everything works fine. That means that all user-written functions that e.g. pattern match on `Bool` Just Work, instead of us having to awkwardly compile the pattern match specially with `ifThenElse`. More special cases, in other words.

On the other hand, `Integer` and `ByteString` really are special, and not treated as datatypes. Random user-written functions that try to pattern match on them (assuming they can see the constructors) will not work, and that's more-or-less okay.

Reply:
I agree that the current version is simpler than compiling pattern matching via `ifThenElse`, but it's probably less efficient. I thought maybe there was a general pattern here and for different sets of built-in types we should also compile the types differently, but if we add, say, `Vector`, it's going to be as "opaque"/"abstract" as `Integer`. Is `Bool` the only type that we don't want to be abstract? Do I remember it correctly that you also handle `Unit` in the same way? But we don't need to eliminate `unitval`, we only need to ignore it, meaning this can be implemented with an abstract `Unit`, right?

Reply:
Well, `Bool` is only really going to be a builtin for convenience, not for performance or because we need it to be abstract. `Integer`, `ByteString`, `Vector`, these would all have "special" implementations that don't map onto whatever Haskell representation we might have.

`Bool` and `Unit` aren't like that at all. And there's a big advantage to handling them normally: normal pattern matching functions etc. Just Work.

Another minor thing: at the moment you can use `Bool` and `Unit` in a builtin-free Plutus Tx program, which you wouldn't be able to do if the mapped onto the builtins.

> But we don't need to eliminate unitval, we only need to ignore it, meaning this can be implemented with an abstract Unit, right?

You're assuming that nobody matches on `Unit`, which they totally do. Right now that all Just Works like normal, no special cases required.

Reply:
Ok, thank you for the explanation.

Reply:
I'll add some of this to a note.